File: Code/spikesortingVTJason/spikesortingVTJason.py
import numpy as np
import scipy.io
import matplotlib.pyplot as plt
from scipy.signal import butter, filtfilt
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from scipy.spatial.distance import mahalanobis
from scipy.stats import chi

## Load the recording data and select the part of interest
# line 4

mat_data = scipy.io.loadmat('10 min recording1.mat')
recording1 = mat_data['recording1'].flatten()

fsSpikes = 50000
rawsignal = recording1[20 * fsSpikes:260 * fsSpikes]

## Bandpass filter for Spikes and LFP
# line 9

Fc1_spikes, Fc2_spikes = 300, 3000
b_spikes, a_spikes = butter(4, [Fc1_spikes, Fc2_spikes], btype='band', fs=fsSpikes)
spikes = filtfilt(b_spikes, a_spikes, rawsignal) * 1000

Fc1_lfp, Fc2_lfp = 0.5, 300
b_lfp, a_lfp = butter(2, [Fc1_lfp, Fc2_lfp], btype='band', fs=fsSpikes)
LFP = filtfilt(b_lfp, a_lfp, rawsignal) * 1000

# Plot
# line 23

time = np.arange(0, len(spikes)) / fsSpikes

plt.figure(figsize=(10, 6))
plt.subplot(3, 1, 1)
plt.plot(time, rawsignal, linewidth=0.5)
plt.title('Raw Signal')
plt.subplot(3, 1, 2)
plt.plot(time, spikes, linewidth=0.5)
plt.title('Filtered Spikes')
plt.subplot(3, 1, 3)
plt.plot(time, LFP, linewidth=0.5)
plt.title('Filtered LFP')

plt.xlabel('Time (s)')
plt.tight_layout()
# plt.show()

## Detect the spike according to the threshold
# line 33

upper_threshold = -100
threshold = -20

ind_above = np.where(spikes < threshold)[0]
vector = np.zeros_like(spikes)
vector[ind_above] = spikes[ind_above]

above_threshold = ind_above
# 1.5 ms delay in sample numbers
delay = int(fsSpikes / 1000 * 1.5)

end_spike_index = []
n = 1

for ii in range(len(above_threshold) - 1):
    if spikes[above_threshold[ii]] != spikes[above_threshold[ii + 1]] and \
       above_threshold[ii] < above_threshold[ii + 1] - delay:
        end_spike_index.append(above_threshold[ii])
        n += 1

spike_index = []
for ii in range(1, len(end_spike_index)):
    window = spikes[end_spike_index[ii] - delay:end_spike_index[ii] + delay]
    M = np.min(window)
    index = np.argmin(window)
    if M > upper_threshold:
        spike_index.append(end_spike_index[ii] - delay + index)

spike_index = np.array(spike_index)

## Delete the fake oscillation spike
# line 67

deleteEv = []

for i in range(len(spike_index)):
    before = range(spike_index[i] - 51, spike_index[i] - 21, 1)
    after = range(spike_index[i] + 20, spike_index[i] + 50, 1)

    beforeFind = [index for index, val in enumerate(spikes[before]) if val < -10 or val > 25]
    afterFind = [index for index, val in enumerate(spikes[after]) if val < -10 or val > 25]

    if len(beforeFind) > 1 or len(afterFind) > 1:
        deleteEv.append(i)

for i in deleteEv:
    spike_index[i] = 0

spike_index = [element for element in spike_index if element != 0]

## Get the 3 ms spike cutout
# line 82

per = 50
detected_spikes = spike_index
num_spikes = len(detected_spikes)
data = np.zeros((num_spikes, 2 * per))

start = np.zeros(num_spikes, dtype=int)
stop = np.zeros(num_spikes, dtype=int)
starttime = np.zeros(num_spikes, dtype=int)

for i in range(num_spikes):
    start[i] = detected_spikes[i] - per
    stop[i] = start[i] + (2 * per)
    data[i][:] = spikes[start[i]:stop[i]]
    starttime[i] = start[i]

## PCA analysis
# line 95

data = np.array(data)

# Perform PCA on 'data'
pca = PCA(n_components=3)  # Specify the number of components (3 in this case)
# pca = PCA()
coeff = pca.fit_transform(data)
score = pca.components_
ev = pca.explained_variance_

# Calculate the mean of 'data'
mu = np.mean(data, axis=0)

# Select the first 3 principal components
pcadata = coeff[:, :3]

# plot data
time = np.arange(0, 2 * per, 1 / fsSpikes) * 1000

plt.figure()
plt.plot(time[:data.shape[1]], data.T, linewidth = 0.5)
plt.title('Spikes')
plt.xlabel('Time (ms)')
plt.ylabel('Voltage (mV)')
plt.xlim(0, 2 * per * 1000 / fsSpikes)
plt.grid(True)

## line 107

desired_k = 3
kmeans = KMeans(n_clusters=desired_k, init='k-means++', n_init=desired_k + 6, random_state=42)
kmeans.fit(data)
IDX = kmeans.labels_
C = kmeans.cluster_centers_
# Define colors for each cluster
color_cluster = [
    [0.259, 0.62, 0.741],
    [0, 0, 1],
    [1, 0, 1],
    [0.949, 0.498, 0.047],
    [1, 1, 0],
    [0, 1, 1],
    [0.5, 0, 1],
    [0, 0.5, 1],
    [1, 0.5, 0],
    [1, 0, 0.5]
]

## Plot each cluster
plt.figure()
for i in range(desired_k):
    # Select data for the current cluster
    cluster_data = data[IDX == i, :]
#     # Time vector for plotting
    time = np.linspace(0, (2 * per) * 1e3 / fsSpikes, cluster_data.shape[1])
#     # Plot each waveform in the cluster
    for waveform in cluster_data:
        plt.plot(time, waveform, color=color_cluster[i], linewidth=0.5)

plt.title('Sorted Neuron Signals')
plt.xlabel('Time (ms)')
plt.ylabel('Voltage (mV)')

## line 126

for i in range(desired_k):
    plt.figure()
    # Select data for the current cluster
    cluster_data = data[IDX == i, :]
    # Time vector for plotting
    time = np.linspace(0, (2 * per) * 1e3 / fsSpikes, cluster_data.shape[1])
    # Plot each waveform in the cluster
    for waveform in cluster_data:
        plt.plot(time, waveform, color=color_cluster[i], linewidth=0.5)

    # Calculate the mean waveform for the current cluster
    meanofdata = np.mean(cluster_data, axis=0)
    # Plot the mean waveform
    plt.plot(time, meanofdata, 'k', linewidth=1.5)  # 'k' is for black color

    # plt.title(f'Sorted neuron signals cluster={i+1}')  # i+1 to convert from 0-based to 1-based indexing
    plt.xlabel('Time (ms)')
    plt.ylabel('Voltage (µV)')
    plt.ylim([-70, 40])

plt.show()

# line 141

pca = PCA(n_components=3)  # Adjust the number of components as necessary
# pca = PCA()
pca.fit(data)
coeff = pca.components_
score = pca.transform(data)
ev = pca.explained_variance_

## line 147
mu = np.mean(data, axis=0)
marker_neuron = ['o', '*', 's', 'x', 'd', 'p', '+', '.', 'v', '>']

# Plotting PCA results
fig, ax = plt.subplots()

marker_size = 15
neuron = {}
for i in range(desired_k):
    cluster_data = data[IDX == i]
    neuron[i] = (cluster_data - mu) @ coeff.T  # Transform to PCA space

    ax.scatter(-neuron[i][:, 0], -neuron[i][:, 1], marker=marker_neuron[i], color=color_cluster[i], s=marker_size, facecolors='none',linewidths=0.5)

ax.set_xlabel('PC1', fontsize=24)
ax.set_ylabel('PC2', fontsize=24)
ax.spines['bottom'].set_linewidth(2)
ax.spines['left'].set_linewidth(2)
ax.tick_params(axis='both', which='major', labelsize=16, width=2)

plt.xlim(-200, 100)
plt.ylim(-100, 100)
plt.title('PCA Clustering')
plt.show()

# Add plot for cluster indices
plt.scatter(range(len(IDX)), IDX)

# line 168

firstcluster = int(input('Please enter the first wanted cluster: '))
secondcluster = int(input('Please enter the second wanted cluster: '))

firstcluster = firstcluster - 1
secondcluster = secondcluster - 1

spike_index = np.array(detected_spikes)
IDX = np.array(IDX)

plt.figure()
hb1 = plt.subplot(2, 1, 1)
plt.plot(np.arange(0, len(rawsignal))/fsSpikes, rawsignal, linewidth=1.3)
plt.title('Spike Trace')
plt.xlabel('Time (s)')
plt.ylabel('Voltage (μV)')
plt.gca().spines['bottom'].set_linewidth(2)
plt.gca().spines['left'].set_linewidth(2)
plt.gca().tick_params(axis='both', which='major', labelsize=16)

timecluster_first = [index + 30 for index, cluster in zip(spike_index, IDX) if cluster == firstcluster]
timecluster_second = [index + 30 for index, cluster in zip(spike_index, IDX) if cluster == secondcluster]

timecluster_first = np.array(timecluster_first)
timecluster_second = np.array(timecluster_second)

plt.scatter(timecluster_first/fsSpikes, -900*np.ones(len(timecluster_first)), s=5, marker='^',
            facecolors=color_cluster[firstcluster], edgecolors=color_cluster[firstcluster], linewidth=0.5)
plt.scatter(timecluster_second/fsSpikes, -900*np.ones(len(timecluster_second)), s=5, marker='^',
            facecolors=color_cluster[secondcluster], edgecolors=color_cluster[secondcluster], linewidth=0.5)

hb2 = plt.subplot(2, 1, 2, sharex=hb1)
plt.plot(np.arange(0, len(spikes))/fsSpikes, spikes, linewidth=1.5)
plt.gca().spines['bottom'].set_linewidth(1.5)
plt.gca().spines['left'].set_linewidth(1.5)

timecluster_first = [index for index, cluster in zip(spike_index, IDX) if cluster == firstcluster]
timecluster_second = [index for index, cluster in zip(spike_index, IDX) if cluster == secondcluster]

timecluster_first = np.array(timecluster_first)
timecluster_second = np.array(timecluster_second)

plt.scatter(timecluster_first/fsSpikes, -60*np.ones(len(timecluster_first)), s=5, marker='^',
            facecolors=color_cluster[firstcluster], edgecolors=color_cluster[firstcluster], linewidth=0.5)
plt.scatter(timecluster_second/fsSpikes, -60*np.ones(len(timecluster_second)), s=5, marker='^',
            facecolors=color_cluster[secondcluster], edgecolors=color_cluster[secondcluster], linewidth=0.5)

plt.figure()
time = np.arange(1/fsSpikes, 60, 1/fsSpikes)
trace = spikes[0:60*fsSpikes-1]
plt.plot(time, trace, linewidth=1.5)
plt.ylabel('Voltage (μV)')

plt.scatter(timecluster_first/fsSpikes, -60 * np.ones(len(timecluster_first)), s=5, marker='^',
            facecolors=color_cluster[0], edgecolors=color_cluster[0])
plt.scatter(timecluster_second/fsSpikes, -60 * np.ones(len(timecluster_second)), s=5, marker='^',
            facecolors=color_cluster[3], edgecolors=color_cluster[3])

plt.xlabel('Time (s)')
plt.ylabel('Voltage (μV)')

# Plot PCA results for selected clusters
plt.figure()
plt.scatter(neuron[firstcluster][:, 0], neuron[firstcluster][:, 2],
            color=color_cluster[firstcluster], marker=marker_neuron[firstcluster], s=25)
plt.scatter(neuron[secondcluster][:, 0], neuron[secondcluster][:, 2],
            color=color_cluster[secondcluster], marker=marker_neuron[secondcluster], s=25)
plt.xlabel('PC1', fontsize=24)
plt.ylabel('PC2', fontsize=24)
plt.gca().spines['bottom'].set_linewidth(1.5)
plt.gca().spines['left'].set_linewidth(1.5)
plt.show()

# Subplots
hc = [None, None, None]

plt.figure()
# First subplot
hc[0] = plt.subplot(3, 1, 1)
plt.plot(np.arange(0, len(rawsignal))/fsSpikes, rawsignal, linewidth=1.5)
timecluster_first = spike_index[IDX == firstcluster] + 30
timecluster_second = spike_index[IDX == secondcluster] + 30
plt.scatter(timecluster_first/fsSpikes, -800*np.ones(len(timecluster_first)),
            s=25, marker='^', facecolors=color_cluster[firstcluster],
            edgecolors=color_cluster[firstcluster])
plt.scatter(timecluster_second/fsSpikes, -800*np.ones(len(timecluster_second)),
            s=25, marker='^', facecolors=color_cluster[secondcluster],
            edgecolors=color_cluster[secondcluster])
plt.xlim(39.8, 40.2)
plt.ylim(-800, 200)
plt.xlabel('Time (s)')
plt.ylabel('Voltage (μV)')

# Second subplot
hc[1] = plt.subplot(3, 1, 2, sharex=hc[0])
plt.xlabel('Time (s)')
plt.ylabel('Voltage (μV)')
plt.title('LFP Signal')

plt.plot(np.arange(0, len(LFP))/fsSpikes, LFP, linewidth=1)
plt.ylim(-600, 400)

# Third subplot
hc[2] = plt.subplot(3, 1, 3, sharex=hc[0])
plt.plot(np.arange(0, len(spikes))/fsSpikes, spikes, linewidth=1)
plt.scatter(timecluster_first/fsSpikes, -100*np.ones(len(timecluster_first)),
            s=25, marker='^', facecolors=color_cluster[firstcluster],
            edgecolors=color_cluster[firstcluster])
plt.scatter(timecluster_second/fsSpikes, -100*np.ones(len(timecluster_second)),
            s=25, marker='^', facecolors=color_cluster[secondcluster],
            edgecolors=color_cluster[secondcluster])

plt.xlabel('Time (s)')
plt.ylabel('Voltage (μV)')
plt.title('Filtered Spikes and Cluster Points')
plt.subplots_adjust(hspace=1)
plt.show()

# Link axes
plt.setp([a.get_xticklabels() for a in hc[:-1]], visible=False)
plt.setp([a.get_yticklabels() for a in hc[:-1]], visible=False)

# Mahalanobis distance and L-ratio calculation
clusterone = neuron[firstcluster][:, :3]
clustertwo = neuron[secondcluster][:, :3]
cov_matrix = np.cov(clusterone.T)
inv_cov_matrix = np.linalg.inv(cov_matrix)

d2 = np.array([mahalanobis(point, clusterone.mean(axis=0), inv_cov_matrix) for point in clustertwo])
MD = np.sort(d2)
ID = MD[min(len(clusterone), len(clustertwo)) - 1]
p = chi2.cdf(MD, df=3)
Lratio = np.sum(1 - p) / len(clusterone)

print("L-ratio:", Lratio)

File: Code/spikeMatlab/spikesortingVTJason.m
%% Load the recording data and select the part of interest
clear
close("all")
load('../spikesortingVTJason/10 min recording1.mat')
fsSpikes=50000;
rawsignal =recording1(20*fsSpikes:260*fsSpikes);

%% Bandpass filter for Spikes and LFP
Fc1=300;
Fc2=3000;
H1 = designfilt('bandpassiir','FilterOrder',4, ...
         'HalfPowerFrequency1',Fc1,'HalfPowerFrequency2',Fc2, ...
         'SampleRate',fsSpikes,'DesignMethod','butter');
spikes=filtfilt(H1,rawsignal)*1000;

Fc1=0.5;
Fc2=300;
H2 = designfilt('bandpassiir','FilterOrder',4, ...
         'HalfPowerFrequency1',Fc1,'HalfPowerFrequency2',Fc2, ...
         'SampleRate',fsSpikes,'DesignMethod','butter');
LFP=filtfilt(H2,rawsignal)*1000;

figure
ha(1)=subplot(3,1,1);
plot(0:1/fsSpikes:(length(spikes)-1)/fsSpikes,rawsignal);
ha(2)=subplot(3,1,2);
plot(0:1/fsSpikes:(length(spikes)-1)/fsSpikes,spikes);
ha(3)=subplot(3,1,3);
plot(0:1/fsSpikes:(length(spikes)-1)/fsSpikes,LFP); %plot the overall spikes
linkaxes([ha(1),ha(2),ha(3)], 'x');

%% Detect the spike according to the threshold
upperthreshold=-100;
threshold=-20;
ind_above=find(spikes<threshold); %find the index where the voltage is bigger than threshold
vector=0*spikes;
for ii=1:length(ind_above)
    vector(ind_above(ii))=spikes(ind_above(ii));
end
above_threshold=ind_above;
delay=fsSpikes/1000*1.5; %1.5ms in length
spike_index=[];
n=1;
spikenum=0;

% find out when the spikes happe n
for ii=1:length(above_threshold)-1 % for all time stamp that has the voltage bigger than threshold
    if (spikes(above_threshold(ii))~=spikes(above_threshold(ii)+1)) && (above_threshold(ii)<above_threshold(ii+1)-delay)
        %the voltage is at its maximum and no spikes for at least 1.5 ms
            end_spike_index(n)=above_threshold(ii);
            n=n+1;
    end
end

spike_index=[];
for ii=2:length(end_spike_index)
        [M,index]=min(spikes(end_spike_index(ii)-delay:end_spike_index(ii)+delay));
        if M>upperthreshold
            spike_index(end+1)=end_spike_index(ii)-delay+1+index;
        end

end

%% Delete the fake oscillation spike
deleteEv = [];
for i = 1:length(spike_index)
    before = spike_index(i)-50:1:spike_index(i)-20;
    after = spike_index(i)+20:1:spike_index(i)+50;
    beforeFind = [find(spikes(before) < -10); find(spikes(before) > 25)];
    afterFind = [find(spikes(after)< -10); find(spikes(after) > 25)];
    if length(beforeFind) > 1 || length(afterFind) > 1
        deleteEv(end + 1) = i;
    end
end

for i = deleteEv
    spike_index(i) = 0;
end
spike_index = nonzeros(spike_index);

%% Get the 3 ms spike cutout
per=50;
detected_spikes=spike_index;
num_spikes=length(detected_spikes);
data=zeros(num_spikes,per);

for i=1:num_spikes
    start(i)=detected_spikes(i)-per;
    stop(i)=start(i)+(2*per);
    data(i,1:2*per+1)=spikes(start(i):stop(i));
    starttime(i)=start(i);
end

%% Do PCA analysis on the spike array
[coeff,score,ev]  = pca(data);
mu=mean(data);
pcadataready=data*coeff;
pcadata=pcadataready(:,1:3);

figure
plot(0:(1/fsSpikes)*1e3:(2*per)*1e3/fsSpikes,data')
title('Spikes')
xlabel('Time(ms)')
ylabel('Voltage(mV)')
set(gca,'LineWidth',2,'FontSize',16,'Fontname','Arial Bold')

desired_k=3;%%cluster number
[IDX,C]=kmeans(data,desired_k,'Distance','cityblock','Display','final','Replicates',desired_k+6);
% [s,h] = silhouette(data,IDX);

figure

cluster=cell(desired_k,1);
color_cluster = {[0.259, 0.62, 0.741],[0 0 1],[1 0 1],[0.949, 0.498, 0.047],[1 1 0],[0 1 1],[0.5 0 1],[0 0.5 1],[1 0.5 0],[1 0 0.5]};
color_centroid = {'r*','gs','bo','kd'};
for i=1:desired_k
    cluster{i}=data(IDX==i,:);
    plot(0:(1/fsSpikes)*1e3:(2*per)*1e3/fsSpikes,cluster{i}','Color',color_cluster{i})
    title('sorted neuron signals')
    xlabel('time (ms)');
    ylabel('voltage (mV)');
    set(gca,'LineWidth',2,'FontSize',16,'Fontname','SansSerif')
    hold on
end

for i=1:desired_k
   figure
    cluster{i}=data(IDX==i,:);
    plot(0:(1/fsSpikes)*1e3:(2*per)*1e3/fsSpikes,cluster{i}','Color',color_cluster{i})
    meanofdata=mean(cluster{i},1);
    hold on
    plot(0:(1/fsSpikes)*1e3:(2*per)*1e3/fsSpikes, meanofdata,'k','LineWidth',1.5)
%     title(['sorted neuron signals cluster=' num2str(i)])
    xlabel('Time (ms)');
    ylabel('Voltage (\muV)');
    ylim([-70 40]);
    set(gca,'LineWidth',1.5,'FontSize',20,'Fontname','SansSerif')

end

[coeff,score,ev]  = pca(data);

neuron=cell(desired_k,1);
mu=mean(data);
marker_neuron = {'o','*','s','x','d','p','+','.','v','>'};
figure
for i=1:desired_k
    sizeofcluster=size(cluster{i});

    neuron{i}=(cluster{i}-repmat(mu,sizeofcluster(1),1))*coeff;

plot(neuron{i}(:,1),neuron{i}(:,2),marker_neuron{i},'Color',color_cluster{i})
% title('PCA analysis')
xlabel('PC1','FontSize',24,'LineWidth',5) % x-axis label
 ylabel('PC2','FontSize',24,'LineWidth',5) % y-axis label

set(gca,'LineWidth',2,'FontSize',16,'Fontname','Arial Bold')

hold on
end

figure
scatter(1:length(IDX),IDX)

firstcluster=input('please enter the first wanted cluster:');
secondcluster=input('please enter the second wanted cluster.:');

figure
hb(1)=subplot(2,1,1);
plot(0:1/fsSpikes:(length(rawsignal)-1)/fsSpikes,rawsignal,'LineWidth',1.3)
hold on
i=firstcluster;
    timecluster{i}=detected_spikes(IDX==i)+30;
    scatter(timecluster{i}/fsSpikes,-900*ones(1,length(timecluster{i})),5,'^','MarkerFaceColor',color_cluster{i},'MarkerEdgeColor',color_cluster{i})
    hold on
    i=secondcluster;
    timecluster{i}=detected_spikes(IDX==i)+30;
    scatter(timecluster{i}/fsSpikes,-900*ones(1,length(timecluster{i})),5,'^','MarkerFaceColor',color_cluster{i},'MarkerEdgeColor',color_cluster{i})

title('Spike Trace')
xlabel('Time(s)')
ylabel('Voltage(\muV)')
set(gca,'LineWidth',2,'FontSize',16,'Fontname','SansSerif')

hb(2)=subplot(2,1,2);
plot(0:1/fsSpikes:(length(rawsignal)-1)/fsSpikes,spikes,'LineWidth',1.5)
set(gca,'FontSize',16, 'FontName','SansSerif', 'Linewidth', 1.5);
hold on
i=firstcluster;
timecluster{i}=detected_spikes(IDX==i);
scatter(timecluster{i}/fsSpikes,-60*ones(1,length(timecluster{i})),5,'^','MarkerFaceColor',color_cluster{i},'MarkerEdgeColor',color_cluster{i})
hold on
i=secondcluster;
timecluster{i}=detected_spikes(IDX==i);
scatter(timecluster{i}/fsSpikes,-60*ones(1,length(timecluster{i})),5,'^','MarkerFaceColor',color_cluster{i},'MarkerEdgeColor',color_cluster{i})

linkaxes([hb(1),hb(2)], 'x');

figure
time = 1/fsSpikes:1/fsSpikes:60;
trace = spikes(1:60*fsSpikes);
length(time)
length(trace)
plot(time,trace,'LineWidth',1.5)
set(gca,'FontSize',16, 'FontName','SansSerif', 'Linewidth', 1.5);
hold on
timecluster{i}=detected_spikes(IDX==firstcluster);
scatter(timecluster{i}/fsSpikes,-60*ones(1,length(timecluster{i})),5,'^','MarkerFaceColor',color_cluster{1},'MarkerEdgeColor',color_cluster{1})
hold on
i=secondcluster;
timecluster{i}=detected_spikes(IDX==i);
scatter(timecluster{i}/fsSpikes,-60*ones(1,length(timecluster{i})),5,'^','MarkerFaceColor',color_cluster{4},'MarkerEdgeColor',color_cluster{4})

figure
i=firstcluster;
plot(neuron{i}(:,1),neuron{i}(:,2),marker_neuron{i},'Color',[0.259, 0.62, 0.741], "Marker", "o")
hold on
i=secondcluster;
plot(neuron{i}(:,1),neuron{i}(:,2),marker_neuron{i},'Color',[0.949, 0.498, 0.047], 'Marker', "square")
xlabel('PC1','FontSize',24,'LineWidth',5) % x-axis label
ylabel('PC2','FontSize',24,'LineWidth',5) % y-axis label
set(gca,'FontSize',16, 'FontName','SansSerif', 'Linewidth', 1.5)

figure
hc(1)=subplot(3,1,1);
plot(0:1/fsSpikes:(length(rawsignal)-1)/fsSpikes,rawsignal,'LineWidth',1.5)
hold on
i=firstcluster;
timecluster{i}=detected_spikes(IDX==i)+30;
scatter(timecluster{i}/fsSpikes,-800*ones(1,length(timecluster{i})),25,'^','MarkerFaceColor',color_cluster{i},'MarkerEdgeColor',color_cluster{i})
hold on
i=secondcluster;
timecluster{i}=detected_spikes(IDX==i)+30;
scatter(timecluster{i}/fsSpikes,-800*ones(1,length(timecluster{i})),25,'^','MarkerFaceColor',color_cluster{i},'MarkerEdgeColor',color_cluster{i})
xlim([39.8,40.2]);
ylim([-800,200]);
%     axis off
% xlabel('Time(s)')
% ylabel('Voltage(\muV)')
% set(gca,'LineWidth',2,'FontSize',16,'Fontname','Arial Bold')
hc(2)=subplot(3,1,2);
plot(0:1/fsSpikes:(length(spikes)-1)/fsSpikes,LFP,'LineWidth',1.5);
ylim([-600,400]);
% axis off
hc(3)=subplot(3,1,3);
plot(0:1/fsSpikes:(length(rawsignal)-1)/fsSpikes,spikes,'LineWidth',1.5)
hold on
i=firstcluster;
timecluster{i}=detected_spikes(IDX==i);
scatter(timecluster{i}/fsSpikes,-100*ones(1,length(timecluster{i})),25,'^','MarkerFaceColor',color_cluster{i},'MarkerEdgeColor',color_cluster{i})
hold on
i=secondcluster;
timecluster{i}=detected_spikes(IDX==i);
scatter(timecluster{i}/fsSpikes,-100*ones(1,length(timecluster{i})),25,'^','MarkerFaceColor',color_cluster{i},'MarkerEdgeColor',color_cluster{i})

linkaxes([hc(1),hc(2),hc(3)], 'x');
xlim([39.8,40.2]);
% axis off
clusterone=neuron{firstcluster}(:,1:3);
clustertwo=neuron{secondcluster}(:,1:3);
d2 = mahal(clustertwo,clusterone);
MD=sort(d2);
ID=MD(min(length(clusterone(:,1)),length(clustertwo(:,2))));
p = chi2cdf(MD,3);
Lratio=sum(1.-p)/length(clusterone(:,1));
